crates/driver/src/domain/competition/mod.rs

```diff
@@ -328,6 +302,127 @@ impl Competition {
     }
 }
 
+/// Encode the solutions into settlements.
+async fn encode_solutions(
+    auction: Auction,
+    eth: Ethereum,
+    simulator: Simulator,
+    solutions: Vec<Solution>,
+) -> Vec<(solution::Id, Result<Settlement, solution::Error>)> {
+    let (sender, mut receiver) = tokio::sync::mpsc::unbounded_channel();
+    let auction = Arc::new(auction);
+    let eth = Arc::new(eth);
+    let simulator = Arc::new(simulator);
+
+    let futures = solutions
+        .into_iter()
+        .map(|solution| {
+            let sender = sender.clone();
+            let auction = auction.clone();
+            let eth = eth.clone();
+            let simulator = simulator.clone();
+            async move {
+                let id = solution.id();
+                observe::encoding(id);
+                let settlement = solution.encode(&auction, &eth, &simulator).await;
+                let _ = sender.send((id, settlement));
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let deadline = auction.deadline().remaining().unwrap_or_default();
+    if tokio::time::timeout(deadline, tokio::spawn(join_all(futures)))
+        .await
+        .is_err()
+    {
+        tracing::warn!("reached timeout while encoding");
+    }
+
+    let mut settlements = vec![];
+    while let Ok(settlement) = receiver.try_recv() {
+        settlements.push(settlement);
+    }
+    settlements
+}
+
+/// Tries to merge multiple settlements into a single settlement.
+/// If merged, the merged settlement is added to the `settlements` vector and
+/// the contained settlements are removed from `settlements` vector.
+/// If the deadline is reached, current state of `settlements` vector is
+/// returned.
+async fn merge_settlements(
+    settlements: &mut Vec<Settlement>,
+    eth: &Ethereum,
+    simulator: &Simulator,
+    deadline: auction::Deadline,
+) {
+    if settlements.len() <= 1 {
+        // Nothing to merge.
+        return;
+    }
+
+    // TODO(#1483): parallelize this
+    // TODO(#1480): more optimal approach for settlement merging
+
+    // Merge the settlements in random order.
+    settlements.shuffle(&mut rand::thread_rng());
+
+    // The merging algorithm works as follows: the [`inner_settlements`] vector
+    // keeps the "most merged" settlements until they can't be merged anymore,
+    // at which point they are moved into the [`settlements`] vector.
+
+    let task = async {
+        // work on the copy to make sure the timeout doesn't leave original in an
+        // inconsistent state
+        let mut inner_settlements = settlements.clone();
+
+        while let Some(settlement) = inner_settlements.pop() {
```

Comment:
Could we instead have the previous step (settlement encoding) return a stream of settlements, and for each new settlement check, if it can be merged into an existing settlement? This way we could combine the two steps of the pipeline and find merged solutions even if the encoding step itself would take longer than the deadline.